### ISSUE TITLE
SHAP library does not support TransformedTargetRegressor, how to calculate SAHP in native units with target transform

### ISSUE THREAD

**Issue by valentasgruzauskas** [shap/shap#2719]
We are working with a regression problem, where we want to apply log transform to target variable. After model training, we want to obtain SHAP values in native units, that they would be suitable for end-user.

When trying to create an explainer by using SHAP library, we receive the folowing error:
`shap.utils._exceptions.InvalidModelError: Model type not yet supported by TreeExplainer: <class 'sklearn.compose._target.TransformedTargetRegressor'>`

I can manualy apply reverse transformation to obtain final prediction correctly, however, I want to receive each SHAP value for individual feature in native units. How could I do it?

```
LightGBM prediction with target transform: [24.23973864]
SHAP explainer estimated value with manual reverse_tranform of final value: 24.23973864452016
SHAP explainer estimated value with manual reverse_tranform of intermediate values: 21.04059855102976
```

Reproducable example:
```
import numpy as np
from lightgbm import LGBMRegressor
import shap
from sklearn.pipeline import make_pipeline
from sklearn.compose import TransformedTargetRegressor

#get data set
X, y = shap.datasets.boston()

#Create model with target transform
pipeline = make_pipeline(
    LGBMRegressor()
)

model_with_target_transform = TransformedTargetRegressor(regressor=pipeline, func=np.log1p, inverse_func=np.expm1)
model_with_target_transform.fit(X, y)

#Create model without target transform, but with transform
y = np.log1p(y)
model_without_target_trasnsform = LGBMRegressor()
model_without_target_trasnsform.fit(X, y)

#Select observation for testing purpose
test = X.iloc[:1]

# Calculate prediction with target transform
lightgbm_prediction_with_target_transform = model_with_target_transform.predict(test)

# Create explainer
# Explainer with target transform is not supported
# explainer = shap.TreeExplainer(model_with_target_transform)
# shap.utils._exceptions.InvalidModelError: Model type not yet supported by TreeExplainer: <class 'sklearn.compose._target.TransformedTargetRegressor'>

explainer = shap.TreeExplainer(model_without_target_trasnsform)
explained = explainer(test)

# Manualy inverse transform shap values
shap_prediction_all_transform = np.expm1(explained.base_values[0] + sum(explained.values[0]))
shap_prediction_seperate_transform = np.expm1(explained.base_values[0]) + sum(np.expm1(explained.values[0]))

print(f"LightGBM prediction with target transform: {lightgbm_prediction_with_target_transform}")

# Manual conversion of shap values to estimation, reverse transformation of shap values + base values is correct.
# Seperate values reverse transformation and than sum is incorrect
print(f"SHAP explainer estimated value with manual reverse_tranform of final value: {shap_prediction_all_transform}")
print(f"SHAP explainer estimated value with manual reverse_tranform of intermediate values: {shap_prediction_seperate_transform}")
```

**Comment by pansism** [comment]
Hi @valentasgruzauskas,

You can approach this problem like this:

First calculate the SHAP values by providing the trained regressor to the `TreeExplainer`:

```python
trained_regressor = model_with_target_transform.regressor_   # Mind the underscore

explainer = shap.TreeExplainer(trained_regressor)
shap_values = explainer.shap_values(X)
```

and then transform them back to the original space, using the trained transformer's `inverse_transform()` method:

```python
trained_transformer = model_with_target_transform.transformer_     # Mind the underscore

def inverse_transformation(x):
    """Apply the inverse transformation to a single element."""
    return trained_transformer.inverse_transform(np.array([x]).reshape(1,-1))

vectorized_func = np.vectorize(inverse_transformation)
shap_values_native = vectorized_func(shap_values)
```

Depending on the size and shape of `shap_values`, the application of `inverse_transform()` can be done more efficiently.

**Comment by ssbsjv** [comment]
I don't think this is correct, @pansism. Consider the case when using a log transform. The inverse_transformation will then simply be np.exp().  When applying the exponential to get the SHAP-values back to the original format these will no longer have the property that they for each observation sum up to the difference between the mean and the estimate. Thus the exponentiated SHAP-values are not really helpful to garner an understanding of the contribution of the variables.

**Comment by GZuin** [comment]
In case someone comes after this later. A solution is to envelop the whole pipeline in a dummy function, and then ask KernelExplainer to explain that.

Something like this:
```
 def dummy_fun(X):
      y_pred = model.predict(X)  # model needs to be declared globaly
      return exp(y_pred) #  You make your transformation back to native units inside the function
      
 explainer = shap.kernel_explainer(dummy_fun, X_sample)
 shap_values = explainer.shap_values(X)
```

It will be slow and an approximation of the true shap values (due to KernelExplainer), but it gets the job done

**Comment by samuelefiorini** [comment]
Hey everyone, jumping back into this discussion.

I wanted to share a workaround I've been using for explaining `TransformedTargetRegressor` models. It helps get around the slow performance of using explainers other than `shap.TreeExplainer`.

I ran into this problem trying to get SHAP values for a `HistGradientBoostingRegressor` wrapped in a `TransformedTargetRegressor` that was using a Yeo-Johnson `PowerTransform`.

Just applying the inverse transform to the SHAP values doesn't really work because it messes up the additivity property and skews the results for each feature.


My workaround is to use a simple first-order Taylor expansion to map the SHAP values from the transformed scale back to the original one. Basically, I scale the SHAP values using the derivative of the inverse transform, calculated at the SHAP base value ($E[z]$). This keeps the additivity property (mostly) intact and makes the explanations match up with the model's actual predictions.

Here's the math breakdown:

if $z$ is the transformed target and $y = g(z)$ is the original target (where $g$ is the inverse transform), the SHAP decomposition is:

$$\hat{z} = E[z] + \sum_j \phi_j^{(z)}$$

If we do a first-order Taylor expansion of $g(\hat{z})$ around the base value $E[z]$, we get:

$$\hat{y} = g(\hat{z}) \approx g(E[z]) + g'(E[z])(\hat{z} - E[z])$$

Plugging the SHAP decomposition back in, we get:

$$\hat{y} \approx g(E[z]) + g'(E[z]) \sum_j \phi_j^{(z)}$$

So, we can just define a new base value $E[y] = g(E[z])$ and transform each SHAP value like this: $\phi_j^{(y)} \approx g'(E[z]) \phi_j^{(z)}$.

The derivative $g'(E[z])$ just acts as a scaling factor.

You can find the code snippet in this Gist: [gist.github.com/samuelefiorini/65cbd5d4c3e58daff8aa621a0a8c07f7](https://gist.github.com/samuelefiorini/65cbd5d4c3e58daff8aa621a0a8c07f7)

It seems to be a decent way to handle these transformed target models. Let me know what you all think.
